### PR TITLE
COLIBRI RACE Support pull request

### PR DESCRIPTION
**Pull request by larryho5**
Please let me know if there has any problem

### CODE REVIEW

File: src/main/drivers/accgyro_spi_mpu6500.c [hunk 1]
```diff
@@ -99,10 +99,14 @@ static bool mpu6500Detect(void)
     uint8_t tmp;
 
     mpu6500ReadRegister(MPU6500_RA_WHOAMI, &tmp, 1);
+#ifdef COLIBRI_RACE
```
Comment (leothehuman) [inline src/main/drivers/accgyro_spi_mpu6500.c]: Why exactly this hack is necessary?

File: src/main/config/config.c [hunk 2]
```diff
@@ -360,9 +360,12 @@ static void resetConf(void)
     masterConfig.version = EEPROM_CONF_VERSION;
     masterConfig.mixerMode = MIXER_QUADX;
     featureClearAll();
-#if defined(CJMCU) || defined(SPARKY)
+#if defined(CJMCU) || defined(SPARKY) || defined(COLIBRI_RACE)
     featureSet(FEATURE_RX_PPM);
 #endif
+//#if defined(COLIBRI_RACE)
```
Comment (leothehuman) [inline src/main/config/config.c]: I don't think commented out code belongs in the repository

File: src/main/io/gps.c [hunk 3]
```diff
@@ -252,11 +254,50 @@ void gpsInit(serialConfig_t *initialSerialConfig, gpsConfig_t *initialGpsConfig)
 
 void gpsInitNmea(void)
 {
+#ifdef COLIBRI_RACE
+    uint32_t now;
+#endif
+
     switch(gpsData.state) {
         case GPS_INITIALIZING:
+#ifdef COLIBRI_RACE
```
Comment (leothehuman) [inline src/main/io/gps.c]: Can we make this code generic and configurable?

File: src/main/config/config.c (hunk 2 above)
Comment (larryho5) [inline src/main/config/config.c]: Sorry, Forgot to delete the test code

File: src/main/drivers/accgyro_spi_mpu6500.c (hunk 1 above)
Comment (larryho5) [inline src/main/drivers/accgyro_spi_mpu6500.c]: I am not sure why, because if it checking the WHOAMI command will return fail.
That's why i make it return true anyway.

File: src/main/io/gps.c (hunk 3 above)
Comment (larryho5) [inline src/main/io/gps.c]: this code just set the GPS go to 115200bps, maybe we can make it configurable in the cleanflight configurator later on.

File: src/main/drivers/accgyro_spi_mpu6500.c (hunk 1 above)
Comment (hydra) [inline src/main/drivers/accgyro_spi_mpu6500.c]: yes, this needs to be fixed.  Is the MPU hardware you're using returning something different?

File: src/main/config/config.c (hunk 2 above)
Comment (hydra) [inline src/main/config/config.c]: agreed.

File: src/main/drivers/accgyro_spi_mpu6500.c [hunk 4]
```diff
@@ -163,8 +167,32 @@ static void mpu6500AccRead(int16_t *accData)
     accData[Z] = (int16_t)((buf[4] << 8) | buf[5]);
 }
 
+#ifdef COLIBRI_RACE
+static void mpu6500SpiInit(void)
```
Comment (hydra) [inline src/main/drivers/accgyro_spi_mpu6500.c]: This feels like i should be generic too.

Some drivers are passed a structure that defines the pins used by the target hardware, this is a good approach, because the target specific code is kept out of the driver code.

**Comment by hydra** [comment]
There needs to be a Board - Colibri Race.md that goes with this.

Does the Colibri support parallel PWM input?  If not then the airPWM map and multiPWM should be empty.  (The sparky target needs fixing up in that area).

The airPPM map needs to be defined.

The PWM mapping code likely needs some special cases or checking - e.g. when enabling the CHANNEL_FORWARDING feature.

File: src/main/target/COLIBRI_RACE/system_stm32f30x.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  ******************************************************************************
+  * @attention
+  *
+  * <h2><center>&copy; COPYRIGHT 2014 STMicroelectronics</center></h2>
+  *
+  * Licensed under MCD-ST Liberty SW License Agreement V2, (the "License");
+  * You may not use this file except in compliance with the License.
+  * You may obtain a copy of the License at:
+  *
+  *        http://www.st.com/software_license_agreement_liberty_v2
```
Comment (leothehuman) [inline src/main/target/COLIBRI_RACE/system_stm32f30x.h]: I'm not sure this is compatible with GPL

File: src/main/target/COLIBRI_RACE/target.h [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#define TARGET_BOARD_IDENTIFIER "CLBR"
+#define BST_DEVICE_NAME "COLIBRI RACE"
```
Comment (hydra) [inline src/main/target/COLIBRI_RACE/target.h]: What is BST?  Where is this used?  Should 'BST' code be in a different PR?

File: src/main/target/COLIBRI_RACE/target.h [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+#define I2C2_SCL_PIN         GPIO_Pin_9
+#define I2C2_SCL_PIN_SOURCE  GPIO_PinSource9
+#define I2C2_SCL_CLK_SOURCE  RCC_AHBPeriph_GPIOA
+#define I2C2_SDA_GPIO        GPIOA
+#define I2C2_SDA_GPIO_AF     GPIO_AF_4
+#define I2C2_SDA_PIN         GPIO_Pin_10
+#define I2C2_SDA_PIN_SOURCE  GPIO_PinSource10
+#define I2C2_SDA_CLK_SOURCE  RCC_AHBPeriph_GPIOA
+
+#define USE_BST
```
Comment (hydra) [inline src/main/target/COLIBRI_RACE/target.h]: More BST code

File: src/main/drivers/timer.c [hunk 8]
```diff
@@ -89,7 +89,7 @@ const timerHardware_t timerHardware[USABLE_TIMER_CHANNEL_COUNT] = {
 #define TIMER_APB2_PERIPHERALS (RCC_APB2Periph_TIM1 | RCC_APB2Periph_GPIOA | RCC_APB2Periph_GPIOB)
 #endif
 
-#if defined(STM32F3DISCOVERY) && !(defined(CHEBUZZF3))
+#if defined(STM32F3DISCOVERY) && !(defined(CHEBUZZF3)) && !(defined(COLIBRI_RACE))
```
Comment (hydra) [inline src/main/drivers/timer.c]: don't think you need this change.

File: src/main/drivers/timer.c [hunk 9]
```diff
@@ -115,6 +115,30 @@ const timerHardware_t timerHardware[USABLE_TIMER_CHANNEL_COUNT] = {
 
 #endif
 
+#ifdef COLIBRI_RACE
+const timerHardware_t timerHardware[USABLE_TIMER_CHANNEL_COUNT] = {
+    { TIM1,  GPIOA, Pin_8,  TIM_Channel_1, TIM1_CC_IRQn,            1, Mode_AF_PP_PD,   GPIO_PinSource8,  GPIO_AF_6}, // PWM1 - PA8
```
Comment (hydra) [inline src/main/drivers/timer.c]: If this is the PPM input you likely don't need '1' to be set (output enable).

File: src/main/io/gps.c (hunk 3 above)
Comment (hydra) [inline src/main/io/gps.c]: Yes, please remove GPS autobaud configurtation changes from this PR.  Make a new PR for the GPS related changes so we can discuss the intention of the code and implementation.

File: src/main/drivers/accgyro_spi_mpu6500.c (hunk 1 above)
Comment (larryho5) [inline src/main/drivers/accgyro_spi_mpu6500.c]: It should be returning something else, but i'm haven't check it yet.
I have read the datasheet, the WHOAMI for that chip should be right, but still returning failed on me.
Did anyone try this code and sure it is working?

File: src/main/drivers/accgyro_spi_mpu6500.c (hunk 4 above)
Comment (larryho5) [inline src/main/drivers/accgyro_spi_mpu6500.c]: Can we just doing like this for now? Since other driver are doing this also.

Comment (leothehuman) [inline src/main/drivers/accgyro_spi_mpu6500.c]: The problem with this approach is that after the code is added and working no one has an incentive to improve it. Copying bad code just makes more bad code, it doesn't make it better. Your desire to add the board is totally understandable, but the community will benefit more from doing that properly and tries to force you to put more work into it before accepting the burden of supporting that code forever. (:

Comment (larryho5) [inline src/main/drivers/accgyro_spi_mpu6500.c]: But where i should put this into? because i saw the other MPU driver is doing the same thing, that's why i put it there. Could you please show me where i should put this code at?

Comment (borisbstyle) [inline src/main/drivers/accgyro_spi_mpu6500.c]: Should that not be added here?
https://github.com/cleanflight/cleanflight/blob/c142e3dfe988286d37e898940b0f772d8968db79/src/main/sensors/initialisation.c#L227-L244

**Comment by borisbstyle** [comment]
Did anyone tested the MPU6500 with interrupt handling? I think the current driver is missing that part?

File: src/main/drivers/accgyro_spi_mpu6500.c (hunk 4 above)
Comment (larryho5) [inline src/main/drivers/accgyro_spi_mpu6500.c]: thx

**Comment by larryho5** [comment]
Please check, I have fixed most problem you guys has been comment. Please let me know if there has any other problem need to be fix. Thank you guys.

File: src/main/target/COLIBRI_RACE/system_stm32f30x.h (hunk 5 above)
Comment (larryho5) [inline src/main/target/COLIBRI_RACE/system_stm32f30x.h]: This file i just copy from STM32F3DISCOVERY. Ia m not sure anything wrong.

File: src/main/io/gps.c [hunk 10]
```diff
@@ -236,8 +236,6 @@ void gpsInit(serialConfig_t *initialSerialConfig, gpsConfig_t *initialGpsConfig)
 
     portMode_t mode = MODE_RXTX;
     // only RX is needed for NMEA-style GPS
-    if (gpsConfig->provider == GPS_NMEA)
-        mode &= ~MODE_TX;
```
Comment (hydra) [inline src/main/io/gps.c]: @larryho5 Not sure you need this change, can you remove it from the PR?

**Comment by hydra** [comment]
There is a bit of a code smell.  The code in mpu6500SpiInit() should not be target specific.  It should be called by any device that has an mpu6500 connected via SPI.

See src/target/NAZE/hardware_revision.c

I think the appropriate place to put the code current in sensors/initialisation.c is in the accyri_spi_mpu6500.c source and call it from mpu6500Detect() as required.

The PR is close to being mergable, just this issue and the comment I made on a bit of gps code changes before it can be merged.

**Comment by hydra** [comment]
Please also rebase against latest master.

**Comment by larryho5** [comment]
Please check, I have fixed the problem which you has been comment.
But i'm not sure how to rebase to the latest master, can you show me how to do it?
Is it somewhere i can click on the github website? or i need to use the git command to do it? use git rebase? or git cherry-pick? or git merge?

**Comment by hydra** [comment]
@larryho5 Please see 7adfeffafb3c1aa7b3d87f4c122f08cb3ce3f6e7

**Comment by hydra** [comment]
I merged all your commits into one after rebasing against master and then fixed it up a bit.

Tell Trappy he owes me one :smile:
